skmultiflow/classification/trees/hoeffding_adaptive_tree.py

```diff
@@ -0,0 +1,489 @@
+
+            self.estimationErrorWeight.add_element(add)
+            # Detect change with Adwin
+            self.ErrorChange = self.estimationErrorWeight.detected_change()
+
+            if (self.ErrorChange is True and old_error > self.get_error_estimation()):
+                self.ErrorChange = False
+
+            # Update statistics call LearningNodeNBAdaptive
+            super().learn_from_instance(X, y, weight, hat)  # CHECK changed self to super
+
+            # call ActiveLearningNode
+            weight_seen = self.get_weight_seen()
+
+            if weight_seen - self.get_weight_seen_at_last_split_evaluation() >= hat.grace_period:
+                hat._attempt_to_split(self, parent, parent_branch)
+                self.set_weight_seen_at_last_split_evaluation(weight_seen)
+
+        # Override LearningNodeNBAdaptive
+        def get_class_votes(self, X, ht):
+
+            dist = {}
+            prediction_option = ht.leaf_prediction
+
+            # TODO
+            # if prediction_option == MAJORITY_CLASS:  # MC
+            #   dist = self.get_observed_class_distribution()
+            # elif prediction_option == NAIVE_BAYES:  # NB
+            #   dist = do_naive_bayes_prediction(X, self._observed_class_distribution, self._attribute_observers)
+
+            # NBAdaptive
+            if self._mc_correct_weight > self._nb_correct_weight:
+                dist = self.get_observed_class_distribution()
+            else:
+                dist = do_naive_bayes_prediction(X, self._observed_class_distribution, self._attribute_observers)
+
+            dist_sum = sum(dist.values())  # sum all values in dictionary
+
+            if dist_sum * self.get_error_estimation() * self.get_error_estimation() > 0.0:
+                normalize(dist_sum * self.get_error_estimation() * self.get_error_estimation(), dist)
+
+            return dist
+
+        # Override NewNode, New for option votes
+        def filter_instance_to_leaves(self, X, split_parent, parent_branch, update_splitter_counts, found_nodes=None):
+            if found_nodes is None:
+                found_nodes = []
+            found_nodes.append(HoeffdingTree.FoundNode(self, split_parent, parent_branch))
+
+    # =============================================
+    # == Hoeffding Adaptive Tree implementation ===
+    # =============================================
+
+    def __init__(self, *args, **kwargs):
+
+        super(HAT, self).__init__(*args, **kwargs)
+        self._alternate_trees = 0
+        self._pruned_alternate_trees = 0
+        self._switch_alternate_trees = 0
+        self._tree_root = None  # CHECK I not utilize _tree_root of HoeffdingTree. OK
+
+    def reset(self):
+        self._alternate_trees = 0
+        self._pruned_alternate_trees = 0
+        self._switch_alternate_trees = 0
+        self._tree_root = None
+
+    # Override HoeffdingTree/BaseClassifier
+    def partial_fit(self, X, y, classes=None, weight=None):
```

Comment:
Not sure if we need to override this function. Please confirm.

Reply:
In this section we are going to create a _new_ learning_node() of type AdaLearningNode() in the case where _tree_root is None, and call then learn_from_instance(). I should do some tests anyway.

Reply:
Override removed.